// react-aspect-ratio/dist/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _defineProperty2 = require('babel-runtime/helpers/defineProperty');

var _defineProperty3 = _interopRequireDefault(_defineProperty2);

var _assign = require('babel-runtime/core-js/object/assign');

var _assign2 = _interopRequireDefault(_assign);

var _objectWithoutProperties2 = require('babel-runtime/helpers/objectWithoutProperties');

var _objectWithoutProperties3 = _interopRequireDefault(_objectWithoutProperties2);

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var CUSTOM_PROPERTY_NAME = '--aspect-ratio';

var AspectRatio = function (_PureComponent) {
  (0, _inherits3.default)(AspectRatio, _PureComponent);

  function AspectRatio(props) {
    (0, _classCallCheck3.default)(this, AspectRatio);

    var _this = (0, _possibleConstructorReturn3.default)(this, (AspectRatio.__proto__ || (0, _getPrototypeOf2.default)(AspectRatio)).call(this, props));

    _this.updateAspectRatio = function (ratio, nextRatio) {
      if (ratio !== nextRatio) {
        _this.setState({
          ratio: nextRatio
        }, function () {
          if (_this.node) {
            // BC for older version of React https://github.com/facebook/react/issues/6411
            // check if React support custom property AFTER
            var customPropertyValue = _this.node.style.getPropertyValue(CUSTOM_PROPERTY_NAME);
            if (!customPropertyValue) {
              _this.node.style.setProperty('--aspect-ratio', nextRatio);
            }
          }
        });
      }
    };

    _this.node;

    _this.state = {
      ratio: props.ratio
    };
    return _this;
  }

  (0, _createClass3.default)(AspectRatio, [{
    key: 'componentDidMount',
    value: function componentDidMount() {
      this.updateAspectRatio('', this.props.ratio);
    }
  }, {
    key: 'componentWillReceiveProps',
    value: function componentWillReceiveProps(nextProps) {
      this.updateAspectRatio(this.props.ratio, nextProps.ratio);
    }
  }, {
    key: 'render',
    value: function render() {
      var _this2 = this;

      var _props = this.props,
          ratio = _props.ratio,
          style = _props.style,
          others = (0, _objectWithoutProperties3.default)(_props, ['ratio', 'style']); // eslint-disable-line no-unused-vars

      var newStyle = (0, _assign2.default)((0, _defineProperty3.default)({}, CUSTOM_PROPERTY_NAME, this.state.ratio), style);

      return _react2.default.createElement(
        'div',
        (0, _assign2.default)({
          ref: function ref(node) {
            _this2.node = node;
          },
          style: newStyle
        }, others),
        this.props.children
      );
    }
  }]);
  return AspectRatio;
}(_react.PureComponent);

AspectRatio.defaultProps = {
  className: 'react-aspect-ratio-placeholder',
  ratio: '1/1'
};
AspectRatio.propTypes = {
  ratio: _propTypes2.default.string.isRequired,
  children: _propTypes2.default.node
};
exports.default = AspectRatio;
